internal/engine/internal/oobackend/client.go

```diff
@@ -0,0 +1,112 @@
+package oobackend
+
+import (
+	"bytes"
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/ooni/probe-cli/v3/internal/engine/internal/multierror"
+)
+
+// TODO(bassosimone): apply similar changes to sessionresolver. In
+// particular, let's avoid writing state in sessionresolver.Resolver
+// by making the KVStore mandatory, which is a small change.
+
+// Client is a client to speak to the OONI backend. In its
+// default configuration (i.e., with no proxies), this client
+// uses the best strategy to speak to the backend. Make sure
+// to fill the mandatory fields before using a Client.
+type Client struct {
+	// KVStore is the mandatory KVStore from which we read
+	// the state and in which we write the state.
+	KVStore KVStore
+
+	// HTTPClientWithProxy is an optional HTTP client that is
+	// configured to always use a proxy. When this client is
+	// present, we always and only use it, and we do not update
+	// our strategy dynamically. In fact, when this client is
+	// present, it means the user has requested the probe to
+	// use a specific proxy explicitly.
+	HTTPClientWithProxy HTTPClient
+
+	// HTTPClientDefault is the HTTP client we use by default
+	// when HTTPClientWithProxy is not set. If HTTPClientDefault
+	// is not set, we use http.DefaultClient.
+	HTTPClientDefault HTTPClient
```

Comment:
We probably need another HTTP client for http3 here. (Another possibility is that we modify the design of the HTTP client to be able to dynamically change what protocol to use with a context, as opposed to creating a new client for each specific need. This means that we probably need to use a different type of client here?)

Reply:
Also, in this moment we're currently vulnerable to deadlock after the round trip, because reading the body is not bound to the context life cycle. This may be a reason to use a different HTTP client type than the one used by the stdlib?